### PR TITLE
Cache the chosen kernel of operators'.

### PR DESCRIPTION
**Pull request by Xreki**

### CODE REVIEW

**Comment by luotao1** [comment]
I test it on PyramidDnn inference case.
Command: `./paddle/fluid/inference/tests/api/test_analyzer_pyramid_dnn --infer_model=third_party/inference_demo/pyramid_dnn/model/ --infer_data=third_party/inference_demo/pyramid_dnn/data.txt --gtest_filter=Analyzer_Pyramid_DNN.profile --paddle_num_threads=1 --repeat=1000 --batch_size=1`

| | 2450 v2 (ms per sample)  | 2620 v3 (ms per sample) |
|---|---|---|
|before|  0.305032 | 0.228627 |
|after| 0.264652  | 0.206517 |
|speedup| 13%  | 9.6% |

File: paddle/fluid/framework/operator.cc [hunk 1]
```diff
@@ -921,57 +921,74 @@ void OperatorWithKernel::RunImpl(const Scope& scope,
...
   auto* dev_ctx = pool.Get(place);
 
-  // check if op[type] has kernel registered.
-  auto& all_op_kernels = AllOpKernels();
-  auto kernels_iter = all_op_kernels.find(type_);
-  if (kernels_iter == all_op_kernels.end()) {
-    PADDLE_THROW(
-        "There are no kernels which are registered in the %s operator.", type_);
-  }
+  if (!kernel_type_) {
```
Comment (luotao1) [inline paddle/fluid/framework/operator.cc:924]: 924行if内部，可以再封装成一个函数么？

File: paddle/fluid/framework/operator.h [hunk 2]
```diff
@@ -541,6 +541,8 @@ class OperatorWithKernel : public OperatorBase {
 
  protected:
   mutable OpKernelConfigsMap kernel_configs_map_;
+  mutable std::shared_ptr<OpKernelType> kernel_type_;
+  mutable std::shared_ptr<OpKernelFunc> kernel_func_;
```
Comment (luotao1) [inline paddle/fluid/framework/operator.h]: 建议使用unique_ptr

File: paddle/fluid/framework/operator.cc (hunk 1 above)
Comment (Xreki) [reply paddle/fluid/framework/operator.cc:924]: Done.

File: paddle/fluid/framework/operator.h (hunk 2 above)
Comment (Xreki) [reply paddle/fluid/framework/operator.h]: Done.

**Pull request merged**